lib/dao.js

```
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
+  var context = {
+    Model: Model,
+    query: { 'where': where },
+    hookState: hookState,
+    options: options,
+  };
+  Model.notifyObserversOf('access', context, doUpsertWithNPK);
+  function doUpsertWithNPK(err, ctx) {
+    if (err) return cb(err);
+    var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
+    if (!Object.keys(data).length) {
+      process.nextTick(function() {
+        cb('data field cannot be empty');
+      });
+      return cb.promise;
+    }
+    if (connector.upsertWithWhere && !isOriginalQuery) {
+      var context = {
+        Model: Model,
+        where: ctx.query.where,
+        data: data,
+        hookState: hookState,
+        options: options,
+      };
+      Model.notifyObserversOf('before save', context, function(err, ctx) {
+        if (err) return cb(err);
+        data = ctx.data;
+        var update = data;
+        var inst = data;
+        if (!(data instanceof Model)) {
+          inst = new Model(data, { applyDefaultValues: false });
+        }
+        update = inst.toObject(false);
+        Model.applyProperties(update, inst);
+        Model = Model.lookupModel(update);
+        var connector = self.getConnector();
+        var doValidate = undefined;
+        if (options.validate === undefined) {
+          if (Model.settings.validateUpsert === undefined) {
+            if (Model.settings.automaticValidation !== undefined) {
+              doValidate = Model.settings.automaticValidation;
+            }
+          } else {
+            doValidate = Model.settings.validateUpsert;
+          }
+        } else {
+          doValidate = options.validate;
+        }
+        if (doValidate === false) {
+          callConnector();
+        } else {
+          inst.isValid(function(valid) {
+            if (!valid) {
+              console.warn('Ignoring validation errors in updateOrCreate():');
+              console.warn('  %s', new ValidationError(inst).message);
+            }
```

If not valid you are just giving a warning; I believe you need to return an error when validate fails and you need to validate!